lib/gen_stage/dispatchers/broadcast_dispatcher.ex

```diff
@@ -45,36 +47,48 @@ defmodule GenStage.BroadcastDispatcher do
   end
 
   @doc false
-  def subscribe(opts, {pid, ref}, {demands, waiting}) do
+  def subscribe(opts, {pid, ref}, state = {demands, waiting, subscribed_processes}) do
     selector = validate_selector(opts)
-    {:ok, 0, {add_demand(-waiting, pid, ref, selector, demands), waiting}}
+
+    if subscribed?(subscribed_processes, pid) do
+      Logger.error(fn ->
+        "#{inspect(pid)} is already registered with #{inspect(self())}. " <>
+          "This subscription has been discared."
+      end)
+
+      {:ok, 0, state}
```

Comment:
I think the stage will continue to track this... so we probably need to change this to something like `{:error, reason}` and make sure the producer can handle such returns and act accordingly. The consumer in this case should receive a cancelation notification with the proper reason.

Reply:
I tried to fix that with e479d14...891dd3d. I am not sure if the test case in `gen_stage_test` is stable enough, or if the assertions are racy.

Reply:
@josevalim ping.